Improve: add new area on move room if user inputs new name

Delwing opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions

It makes move area dialog combo box editable. If user inputs not existing area it will be created and room will be moved there.

#### Motivation for adding to Mudlet

One step move room to new area.

#### Other info (issues closed, discussion etc)

closes #1855

Delwing committed:
Improve: add new area on move room if what user inputs is not in existing list

Delwing committed:
Merge branch 'development' into add-new-area-on-move

Delwing committed:
remove redundant spaces

Delwing committed:
Merge branch 'add-new-area-on-move' of github.com:Delwing/Mudlet into add-new-area-on-move

Delwing committed:
add message confirmation that area was created

vadi2 reviewed (commented):
Great idea!

Using c4e71b65 I'm not seeing my new area show up in the area list on the mapper and there's no feedback on the action that I just did.

Delwing committed:
remove spaces

Delwing commented:
Are you sure? Just checked and everything works fine for me.
![image](https://user-images.githubusercontent.com/3740628/193614595-03040645-8d84-46b2-9b31-5c948eceb1f2.png)

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4241,6 +4241,11 @@ void T2DMap::slot_setArea()
 
+    connect(arealist_combobox, &QComboBox::currentTextChanged, this, [=](QString newText) {
+        auto buttonBox = set_room_area_dialog->findChild<QDialogButtonBox*>("buttonBox");
+        buttonBox->button(QDialogButtonBox::Ok)->setEnabled(!newText.isEmpty());
```
Is `newText` not being empty a sufficient validation test to enable the "Ok" button?

Delwing replied:
I think so - either someone will pick/type existing area or new area will be created.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4257,13 +4262,29 @@ void T2DMap::slot_setArea()
+        newAreaId = mpMap->mpRoomDB->addArea(newAreaName);
+        if (newAreaId == 0) {
+            return; // This should not happen, but if it fails addArea underneath will log reason
+        }
```
```suggestion
        if (!newAreaId) {
            // This should not happen, but if it fails TRoomDB::addArea(...) above will log a reason to the main console
            mpMap->postMessage(tr("[ ERROR ] - Unable to add \"%1\" as an area to the map.\n"
                                  "See the \"[MAP ERROR:]\" message for the reason.",
                                  // Intentional separator between argument
                                  "The '[MAP ERROR:]' text should be the same as that used for the translation of \"[MAP ERROR:]%1\n\" in the 'TMAP::logerror(...)' function.").arg(newAreaName));
            return;
        }
```
:astonished: I've just checked and the error message that is referred to does not use the same system/formatting/presentation style as the "OK" message does so I think an "ERROR" message inviting the user to see what that says is a little more user-friendly then just the other one being squirted at them... :grinning:

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4257,13 +4262,29 @@ void T2DMap::slot_setArea()
+        if (newAreaId == 0) {
+            return; // This should not happen, but if it fails addArea underneath will log reason
+        }
+        mpMap->postMessage(tr("[  OK  ]  - Added %1 (%2) area to map.").arg(newAreaName, QString::number(newAreaId)));
```
```suggestion
        mpMap->postMessage(tr("[  OK  ]  - Added \"%1\" (%2) area to map.").arg(newAreaName, QString::number(newAreaId)));
```

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4257,13 +4262,29 @@ void T2DMap::slot_setArea()
+            return; // This should not happen, but if it fails addArea underneath will log reason
+        }
+        mpMap->postMessage(tr("[  OK  ]  - Added %1 (%2) area to map.").arg(newAreaName, QString::number(newAreaId)));
+        mpMap->mUnsavedMap = true;
```
:bulb: **NOTE:** this will need revising:
```diff
-        mpMap->mUnsavedMap = true;
+        mpMap->setUnsaved(__func__);
```
when #6285 hits the codebase.

Delwing replied:
Sure thing. I'll make appropriate commit.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4257,13 +4262,29 @@ void T2DMap::slot_setArea()
+
+        if (mpMap->mpMapper) {
+            mpMap->mpMapper->updateAreaComboBox();
+        }
```
The check *seems* unnecessary to me - the `T2DMap` class is used/created *by* what `mpMap->mpMapper` (a `dlgMapper` class instance) is pointing to - and one could not get to this point in the application's execution without it pointing to a valid thing - can it? So I think it should be okay to have:
```suggestion
        mpMap->mpMapper->updateAreaComboBox();
```

SlySven reviewed (commented):
You've made the `arealist_combobox` editable - luckily (bool) `QComboBox::duplicatesEnabled` defaults to `false` - which will help to prevent some issues. I haven't tried this yet so I'll be curious to see what happens if the "Default area" is NOT being shown but the user enters it as a new area name...!

Delwing committed:
Update src/T2DMap.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Delwing committed:
Update src/T2DMap.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Delwing committed:
Update src/T2DMap.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Delwing committed:
Merge branch 'development' into add-new-area-on-move

vadi2 commented:
Here's how it looks like using 3480d1ef -

[Screen recording 2022-10-04 1.37.43 PM.webm](https://user-images.githubusercontent.com/110988/193810381-a15a3d1c-2542-4047-9e3c-d321d286f502.webm)

Delwing committed:
Merge branch 'development' into add-new-area-on-move

Delwing commented:
@vadi2 Can you please check on newer build? I've just checked fresh one, repeating your steps with new map and works correctly.

demonnic commented:
When I try to input a name into the area list it just sent it to an area named 'unnamed' which as it turns out already had some other rooms in it in this map. I had to use centerview() to find it again.

Delwing commented:
@demonnic 'unnamed' with name like that? Or with "empty" name?
Moving rooms never took you to destination area.

demonnic commented:
Just double checked since I seemed to have mixed up versions on another PR but this still sent it to `(unnamed)` , with the parentheses. It is also the top area in the list, so that could be why it was sent there. 

I'm now getting inconsistent results, with it working sometimes and not others. But I'm unable to track down why it's failing the times it is failing. I stopped after it failed the first time when I tested before.

Delwing commented:
Hmm never failed single for me, so not sure what can make it to fail :)

Delwing committed:
clear selection after move

demonnic commented:
So I downloaded the latest windows build, and I moved a room from one area to another existing area, then tried to create a new area via typing in the name in the move dialog and it went to the first one in the list instead of making a new one. Then I moved it back, and redid the exact same process and it created the new area. 

And then it seems to work every time after that.

Delwing committed:
go to target area after move

demonnic commented:
links are out of date, I've kicked off another set of builds to get new ones and will retest later tonight or tomorrow

vadi2 committed:
Merge branch 'development' into add-new-area-on-move

vadi2 committed:
Merge branch 'development' into add-new-area-on-move

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -4076,13 +4081,31 @@ void T2DMap::slot_setArea()
+            return;
+        }
+        mpMap->postMessage(tr("[  OK  ]  - Added \"%1\" (%2) area to map.").arg(newAreaName, QString::number(newAreaId)));
+        mpMap->mUnsavedMap = true;
```
```suggestion
        mpMap->setUnsaved(__func__);
```

vadi2 committed:
Update for new setUnsaved function

SlySven committed:
Merge branch 'development' into add-new-area-on-move

SlySven committed:
Merge branch 'development' into add-new-area-on-move

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4110,9 +4133,12 @@ void T2DMap::slot_setArea()
             }
+            auto& targetAreaName = areaNamesMap.value(newAreaId);
+            mpMap->mpMapper->comboBox_showArea->setCurrentText(targetAreaName);
+            slot_switchArea(targetAreaName);
```
:warning: This code is version sensitive it will only compile with Qt versions ***less than* 5.15.0** as the method is NOT a slot in the later versions!
```suggestion
#if (QT_VERSION) >= (QT_VERSION_CHECK(5, 15, 0))
            switchArea(targetAreaName);
#else
            slot_switchArea(targetAreaName);
#endif
```

SlySven reviewed (changes requested):
You need to revise something!

SlySven commented:
Humm, after locally fixing the thing I mentioned I notice that the behaviour on editing an existing entry is not as I'd expect. If I leave it on the first item in the list and delete the ` (<count>)`  AND one or more characters in the original area name, then click Ok, then *nothing happens*.

This is not what I would expect, which was one of:
* the **selected rooms** get moved to a new area with the modified (shortened name), but other rooms in the same original area are left in the original area.
* all rooms in the original area stay in that area but it's name is modified as per the edit. Note that this MUST NOT happen should any editing be attempted on the `(QString) TMap::mDefaultAreaName` entry (if it *is* visible in that widget).

I can see that one probably suspect thing is the lambda function in `(void) T2DMap::slot_setArea()`:
```cpp
    connect(arealist_combobox, &QComboBox::currentTextChanged, this, [=](QString newText) {
        auto buttonBox = set_room_area_dialog->findChild<QDialogButtonBox*>("buttonBox");
        buttonBox->button(QDialogButtonBox::Ok)->setEnabled(!newText.isEmpty());
    });
```
although this is fired every time a character is changed by editing the text I can see the lambda is primarily associated with disabling the OK button should the text for that entry being an empty string {in fact the check should probably be `!newText.trimmed().isEmpty()` because I think we trim spaces off elsewhere and a `newText` that is just spaces will get rejected later on - or by the area name routines elsewhere}.

 Importantly, later one in that method - after the main event-loop stopping `set_room_area_dialog->exec()` has returned (this is dodgy but we have used it elsewhere even though it could bite us) -  the following:
```cpp
    auto newArea = arealist_combobox->itemData(arealist_combobox->currentIndex());
    int  newAreaId;
    if (newArea.isValid()) {
        newAreaId = newArea.toInt();
    } else if (sortedAreaList.contains(arealist_combobox->currentText().trimmed())) {
        newAreaId = mpMap->mpRoomDB->getAreaNamesMap().key(arealist_combobox->currentText());
    } else {
        auto newAreaName = arealist_combobox->currentText();
        newAreaId = mpMap->mpRoomDB->addArea(newAreaName);
        if (!newAreaId) {
                        mpMap->postMessage(tr("[ ERROR ] - Unable to add \"%1\" as an area to the map.\n"
                                  "See the \"[MAP ERROR:]\" message for the reason.",
                                  // Intentional separator between argument
                                  "The '[MAP ERROR:]' text should be the same as that used for the translation of \"[MAP ERROR:]%1\n\" in the 'TMAP::logerror(...)' function.").arg(newAreaName));
            return;
        }
        mpMap->postMessage(tr("[  OK  ]  - Added \"%1\" (%2) area to map.").arg(newAreaName, QString::number(newAreaId)));
        mpMap->setUnsaved(__func__);

        mpMap->mpMapper->updateAreaComboBox();
    }
```
is not going to edit an existing entry because as far as I can see, `newArea` as retrieved from the `arealist_combobox` current item is going to be the existing entry's area ID and that will be a valid number so the `newAreaId` will be set to that value by the `if` and neither the `else if` or `else` sections following it will be executed so nothing will be changed.

SlySven commented:
:thinking: It is possible that we might need a dummy entry at the start of the `QComboBox` with an entry of something like `(new area)` and a `Qt::UserRole` data entry of `QVarient<int>( 0 )` {`0` has been coded to be an ***invalid*** area ID number BTW} then, if that value is selected the execution *will* get to the `else` part and if the `newAreaName` is NOT that dummy value then it can be tested and if okay can be used as a new area name...

Also, I have been experimenting with the same sort of "magic" code that I put into the `dlgVarsMainArea::dlgVarsMainArea(QWidget* pParentWidget)` constructor to display, but make *disabled* (to allow display but not allow selection of) **some** of the values - though in this I was looking to modify the `Qt::ItemIsEditable` rather than the `Qt::ItemIsEnabled` flag - so that some entries could be selected but NOT modified. This would be applied to the "Default area" name and, I think, any areas when some, but not ALL the rooms from that area are in the selection. The reason for the latter is because I think that if the user selects **some** of the rooms in an area and then ***modifies*** that area name it is unclear whether they would want:
* EITHER: ALL the rooms in the area to be kept in that area, and ANY rooms selected in a DIFFERENT area to be move to THAT area AND the name to be changed to the new value.
* OR: just the selected rooms (from any area) to be moved to a NEW area with the modified name and a new ID but the unselected rooms to be left in the original area.

It isn't obvious but it is possible IIRC (but tricky with careful use of the main mapper area selection widget) to select rooms from more than one area at a time - it is a corner-case but if not handled it might produce confusing results...

I did get close to something last night but had to :sleeping_bed: before I could check that it was working fully - I did however manage to add tool-tips to each of the items in the `QComboBox` that were displayable when the selection was opened-up explaining how many rooms in the selection were from that area - and if they represented ALL the rooms in that area (and/or if it was the "Default area") - which also explained whether that particular area name could be edited or not. I just didn't have the time to see if it worked to do that or not, but that will be later today.

Kebap commented:
This Kind of confusing UI upon multiple selection makes me think this may not be the best place to add and/or rename an area after all. 

Also please beware scope creep. All the additional information shown may create lots of bloat. YAGNI. Tool tips are not very accessible.

SlySven commented:
> This Kind of confusing UI upon multiple selection makes me think this may not be the best place to add and/or rename an area after all.

You think we need an "Area properties" dialogue for the 2D mapper then? :grinning: 

> Also please beware scope creep. All the additional information shown may create lots of bloat. YAGNI. Tool tips are not very accessible.

TBH The tool-tips was just something to help me debug what is going on - I need not include them in a PR to **this** PR - OTOH the corner case I describe is exactly the sort of thing that a supposedly simple "unified" UI chokes on and if we don't explicitly avoid getting into those corner cases (by saying, "Umm, you can't do that now with *this*!") we will only get the end-users moaning when they do something extreme and things go :boom:.

For example: the current "Symbol colour" stuff fails to handle multiple rooms with the same graphemes for the room symbols ***but different foreground colours***...

vadi2 committed:
Update src/T2DMap.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
Yes, we discussed that over there at that time already, and the best possible solution to handle all combinations of corner cases is if you provide the player with a full table of each room and each property so they can freely mix and modify all at once or some or not. With a simplified UI, there will be compromises necessarily.

> You think we need an "Area properties" dialogue for the 2D mapper then? 😀

In fact that was my original line of thought, before this design was suggested here.

Now reading through this PR conversation again, I am not sure where was the notion introduced that the UI should be able to rename an existing area? I think it's not there. Then I don't see how you could arrive at your suggested reaction 2 (i.e. OR: just the selected rooms (from any area) to be moved to a NEW area with the modified name and a new ID but the unselected rooms to be left in the original area.)

I would expect reaction 1, in line with what the PR author originally described:
> If user inputs not existing area it will be created and room will be moved there.

This can include multiple rooms, and multiple rooms from different areas as well. I see no contradiction or ambiguity there at all. Why did you?

Only doubt then again, selecting multiple rooms from differing areas, may create further problems down the line, if the rooms happen to have same coordinates, then in the new area the rooms will overlap. I think it is an OK result to have. We should not further bloat up this PR to accommodate. It can be easily fixed manually eventually.

SlySven commented:
> Now reading through this PR conversation again, I am not sure where was the notion introduced that the UI should be able to rename an existing area? I think it's not there. 

By making the `QComboBox` "**[editable](https://doc.qt.io/qt-6/qcombobox.html#editable-prop)**" it allows the user to modify any *existing* entries AFAICT, unless I am mistaken. :worried: - which is why I was looking into the code that I put into the `dlgVarsMainArea` constructor which manipulates the flags for individual items in the `QComboBox` - in that situation it was disabling some entries so that they could be displayed but not selected whereas here I was looking to prevent some (most) entries from being editable.

Kebap commented:
You can edit text in the popup alright, but do we want area names to be modified? I think not. They count as new area then.

edit: Then again, I do want area names to be editable, but it is not part of the original PR suggested here, if I understand it right. That is why I said an area properties editor may be better in the long run. However that one will not be used to move rooms at the same time, and is not in scope here either.

Let's ask @Delwing for confirmation, maybe?

SlySven commented:
> ... but it is not part of the original PR suggested here ...

:confused: Hang on! That is precisely what **the title of this PR does say** - that when you are moving rooms to another area from the 2D mapper context menu that you CAN provide a new area name and it will get created as a destination for those rooms!

Kebap commented:
Yes, you create a new area.
You do not rename an existing area.
Even though you edit text in a box.
Not sure how to rephrase.

SlySven commented:
> Yes, you create a new area.
> You do not rename an existing area.
> Even though you edit text in a box.
> Not sure how to rephrase.

But it doesn't work like that:

Here I have selected a range of rooms (just to test things they are actually from 2 different areas, the 4000s ones are from "White Tower" ID = 10 and the 1000s ones are from "Amador" ID = 17) - as you can see the dialogue comes up showing the first area (sorted alphabetically):
![Screenshot_20230108_173855](https://user-images.githubusercontent.com/6163092/211210946-f34367f5-bf51-4dd9-b178-ab4f82da0ef7.png)

I then edit that first item shown - **as I can because ALL the entries are now editable because the `QComboBox` has been set to be in that mode**: 
![Screenshot_20230108_173951](https://user-images.githubusercontent.com/6163092/211210944-d8682110-f2ca-4369-bf9a-4248b1885946.png)

Pressing enter makes the box go away :
![Screenshot_20230108_174055](https://user-images.githubusercontent.com/6163092/211210942-c70de25e-50cb-43f4-a8a4-e7e5ac016548.png)

However things have go awry - at least for the rooms (the ones in the 4000s) that were in area 10 initially - a new area has NOT been created and thus none of the rooms were moved to "Amador (new)" but the ones in area 10 ("White Tower") *have* been shifted - but to area 17 "Amador" and not the new area ***as that wasn't created***:
![Screenshot_20230108_174302](https://user-images.githubusercontent.com/6163092/211210939-d3ab090a-e6e7-457e-ac03-7a3be9d6ddd0.png)

![image](https://user-images.githubusercontent.com/6163092/211212198-bfd5b5b0-650f-4dd9-9660-a905d0eedd88.png)

SlySven commented:
What I will try now is to put in a "(new area)" dummy at the top of the list - which **can** be edited and then try and prevent all the other existing ones from being modified. Then if that first one is changed to something else which is a valid new area name - allow that to be created and used. I will also look to see if it is possible to lock the "Ok" until the selection matches one of the other entries or that one is edited to be a valid new one.

Kebap commented:
> However things have go awry

I think it's a bug, because @Delwing did not expect you to select rooms from different areas. 
In fact, the original description here only mentions one room and not multiple.
Please wait for feedback from @Delwing before you add on to their logic.

I don't see how adding a new selection improves anything but makes handling more complicated.

Delwing commented:
I am back from short vacation. I will read through and get back to you tomorrow at latest.

SlySven commented:
> I think it's a bug, because @Delwing did not expect you to select rooms from different areas.

He may not have expected it - but since it is possible you can bet that someone will do it - I'm just pre-empting them!
> There is a race between mankind and the universe. Mankind is trying to build bigger, better, faster, and more foolproof machines. The universe is trying to build bigger, better, and faster fools. So far the universe is winning.
>    ---- Albert Einstein

Delwing commented:
Why on earth would you like to edit area name on "move room" dialog?
This is just modified original move room dialog that allows creation of new area, just as requested in https://github.com/Mudlet/Mudlet/issues/1855

This had some issues initially, that I couldn't trace, but I guess your first commit fixes that. 
After PC update I don't have env for C++ dev yet, so I have to dedicate some time (which I currently don't have much) to prepare it.

vadi2 commented:
If you wanted to move the rooms to an area that doesn't exist yet? I don't see it as such a terrible thing.

Kebap commented:
> If you wanted to move the rooms to an area that doesn't exist yet?

Then a new area will be created alright. That is the gist of this PR.
Only for some reason a notion of renaming existing areas came up here.
I think it is a great misunderstanding and not required or requested at all.

Then again, SlySven uncovered another issue with moving multiple rooms (from differing areas) at the same time.

vadi2 commented:
Agree, renaming existing areas seems too far.

SlySven commented:
> Agree, renaming existing areas seems too far.

TBH I agree - and what I am looking at is to prevent the user getting the impression that they could do so. @Kebap 's use of an editable value at the top of the `QCombBox` items does seem like a viable approach to also use here so I am just working on something like that at the moment (well, technically, it is on my Main machine and I am on my Laptop - in a :coffee: shop at present but I'd hope to have something to show in the next 24 hours)...

Delwing commented:
Maybe we can add small notion when we enter area name that does not exists that area WILL be created.

vadi2 commented:
Agreed. Just as an fyi, not a warning.

SlySven commented:
~~Um, https://github.com/Delwing/Mudlet/pull/81 contains my attempt at "improving" things - I am not sure why it isn't being CI/CBed - perhaps it needs to be merged into @Delwing 's repo before it gets applied here. It may mean that one will have to build it locally to try it out...~~ Closed and branch deleted.

Delwing committed:
Merge branch 'development' into add-new-area-on-move

Delwing committed:
Merge remote-tracking branch 'upstream/development' into add-new-area-on-move

Delwing committed:
fix application crash after move room to area

Delwing committed:
- restore deleteOnClose for set_room_area_dialog
- remove modality of dialog

Delwing committed:
fix checking whether this is new area
add notation when new area will be created

Delwing committed:
resolve conflicts

Delwing committed:
minor code style fixes

Delwing commented:
@vadi2 @SlySven @demonnic 
I think changes I did should satisfy recent conversations + fix previous issue when this was just not working...

I didn't merge SlySven's PR into mine, since I am really not sure if we should add edit area functionality there.

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -4075,15 +4077,46 @@ void T2DMap::slot_setArea()
+        auto buttonBox = set_room_area_dialog->findChild<QDialogButtonBox*>("buttonBox");
+        buttonBox->button(QDialogButtonBox::Ok)->setEnabled(!newText.trimmed().isEmpty());
+        if (!newText.trimmed().isEmpty() && arealist_combobox->findText(newText.trimmed(), Qt::MatchExactly) == -1
+            && !sortedAreaList.contains(newText.trimmed().trimmed())) {
```
```suggestion
            && !sortedAreaList.contains(newText.trimmed())) {
```

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -4075,15 +4077,46 @@ void T2DMap::slot_setArea()
+        buttonBox->button(QDialogButtonBox::Ok)->setEnabled(!newText.trimmed().isEmpty());
+        if (!newText.trimmed().isEmpty() && arealist_combobox->findText(newText.trimmed(), Qt::MatchExactly) == -1
+            && !sortedAreaList.contains(newText.trimmed().trimmed())) {
+            label_info->setText(qsl("This will create new area: %1").arg(arealist_combobox->currentText()));
```
```suggestion
            label_info->setText(tr("This will create new area: %1").arg(arealist_combobox->currentText()));
```

vadi2 reviewed (commented):
Looks good, one point from my side is that compared to the current dialog there's really no indication that you can create new areas now. I think people will not find this feature because nothing indicates that you can create new areas now.

How do you think you could make it discoverable?

Delwing committed:
Apply suggestions from code review

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

Delwing commented:
I guess we can add label there.

Commit:
resolve conflicts

Commit:
add notion that that new area can be crated

Delwing commented:
I've added initial text to info label below "Input new area name to create one.". I've also changed its font size to 10pt, to not steal away focus of user.

vadi2 committed:
Merge branch 'development' into add-new-area-on-move

vadi2 commented:
I like the idea! Though the font size seems to be the same for me:

![image](https://user-images.githubusercontent.com/110988/229703143-7063b9b5-1434-46bd-bb48-445c33acb3b1.png)

You might have to do this dynamically in code as it can differ per OS?

Delwing commented:
I'll check. I did put it as 10 fixed, it is possible that default size is controlled outside of Mudlet.

Commit:
make info label smaller

Delwing commented:
@vadi2 it is dynamic right now.

Commit:
Merge branch 'add-new-area-on-move' of github.com:Delwing/Mudlet into add-new-area-on-move

vadi2 reviewed (approved):
Looks great on Linux now, thanks!

Delwing commented:
@SlySven can you recheck please?

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4097,6 +4097,11 @@ void T2DMap::slot_setArea()
     set_room_area_dialog->setAttribute(Qt::WA_DeleteOnClose);
     arealist_combobox = set_room_area_dialog->findChild<QComboBox*>("arealist_combobox");
+    auto label_info = set_room_area_dialog->findChild<QLabel*>("label_info");
+    auto font = QFont(QApplication::font());
```
IIRC just declaring something as a `QFont` will use the application font anyhow. I *think* this will do the same:
```suggestion
    QFont font;
```
though I guess it might be worth checking before relying on it... :wink:

Delwing replied:
Yep! This is indeed what doc says and it's working as expected that way.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4097,6 +4097,11 @@ void T2DMap::slot_setArea()
     arealist_combobox = set_room_area_dialog->findChild<QComboBox*>("arealist_combobox");
+    auto label_info = set_room_area_dialog->findChild<QLabel*>("label_info");
+    auto font = QFont(QApplication::font());
+    font.setPointSize(font.pointSize() - 1);
```
So, you want the next size down *whatever* it is? :confused:

Delwing replied:
Yep. -1 to default font size.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4097,6 +4097,11 @@ void T2DMap::slot_setArea()
+    auto font = QFont(QApplication::font());
+    font.setPointSize(font.pointSize() - 1);
+    label_info->setFont(font);
+    arealist_combobox->setInsertPolicy(QComboBox::NoInsert);
```
You really want this line ***after*** the check that `arealist_combobox` is not a `nullptr` in the next line - unless the latter is not needed as otherwise things will go :boom: before it does that check!

Delwing replied:
You are totally right, doesn't make sense that way.
I moved null check right after assignment.

SlySven reviewed (approved):
Other than that use-before-nullptr-check I think this is otherwise good to go. Certainly in testing I couldn't break things like before... :grinning: 

I have dis-engaged the auto merge so that my giving approval doesn't trip it into going ahead - fix that issue and you can then go and merge it...

Delwing committed:
move null-check right after assignment

Delwing committed:
change used QFont constructor
